CONTRIBUTING.md

```diff
@@ -252,6 +252,74 @@ There may, at times, need to be exceptions where breaking changes are allowed in
 discretion of the project's maintainers, and must be carefully considered before merging. An example of an allowed
 breaking change might be a fix for a behavioral bug that was released in an initial minor version (such as `v0.3.0`).
 
+
+## API conventions
+
+This project follows the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md). Minor modifications or additions to the conventions are listed below.
+
+### Optional vs. Required
+
+* Status fields MUST be optional. Our controllers are patching selected fields instead of updating the entire status in every reconciliation.
+
+Optional fields have the following properties:
+* An optional field MUST be marked with `+optional` and include an `omitempty` JSON tag.
+* Fields SHOULD be pointers if the nil and the zero values (by Go standards) have semantic differences.
+  * Note: This doesn't apply to map or slice types as they are assignable to `nil`.
+
+#### Example
+ 
+When using ClusterClass, the semantic difference is important when you have a field in a template which will
+have instance-specific different values in derived objects. Because in this case it's possible to set the field to `nil`
+in the template and then the value can be set in derived objects without being overwritten by the cluster topology controller.
+
+#### Exceptions
+
+* Fields in root objects should be kept as scaffolded by kubebuilder, e.g.:
+  ```golang
+  type Machine struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   MachineSpec   `json:"spec,omitempty"`
+    Status MachineStatus `json:"status,omitempty"`
+  }
+  type MachineList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []Machine `json:"items"`
+  }
+  ```
+
+* Top-level fields in `status` must always have the `+optional` annotation. If we want the field to be always visible even if it 
+  has the zero value, it must **not** have the `omitempty` JSON tag, e.g.:
+  * Replica counters like `availableReplicas` in the `MachineDeployment`
+  * Flags expressing progress in the object lifecycle like `infrastructureReady` in `Machine`
+
+### CRD additionalPrinterColumns
+
+All our CRD objects should have the following `additionalPrinterColumns` order (if the respective field exists in the CRD):
+* Namespace (added automatically)
+* Name (added automatically)
+* Cluster
+* Other fields
+* Replica-related fields
```

Comment:
do we may be want to recommend printing status vs intent?

Reply:
Good question.

Context: Afaik we cannot have columns based on multiple values, e.g. Kubernetes Pod 4/5 containers are ready.

 think the only thing we can do is try to avoid misunderstandings if something is spec or status through good column names. Let's see what we currently have:
* I think for a lot of columns it doesn't really matter: e.g.: Cluster, Age, NodeName
* It seems like our replica columns are also easy to understand: Replicas is .spec.Replicas and all others are from status, e.g. Ready, Updated, Unavailable
* Similar for our "object lifecycle flags", like: Initialized, API Server Available

I think the main problem is when, based on the column name, the column could be spec or status and if the status could be different from spec until the object is entirely reconciled. For me that's mostly `Version` (although the MHC fields and columns could also need some improvement => https://github.com/kubernetes-sigs/cluster-api/issues/5291).

Regarding `Version`: One idea to clear this up would be to rename the column in cases where it's not clear to something like `SPEC_VERSION` or `DESIRED_VERSION` (I don't have good ideas :)). I think this would apply to `Cluster` and `MachineDeployment` (Machines only ever have one version and afaik MachineSets to (?)).

But back to topic, I'm not entirely sure what the convention could or should be. I think for me it would be let's try to make clear through the column name if something is spec or status for fields where they could be misunderstandings.

Reply:
I would add the following:
> **NOTE**: In cases where it's not clear if the column field is part of spec or status, we SHOULD try to make it obvious by choosing the column name accordingly.

Regarding `VERSION`. Should we make a change to the `VERSION` column of Cluster and MachineDeployment before the v1beta1 release? As mentioned, my best idea for another name is probably `DESIRED_VERSION` (alternative `SPEC_VERSION`).

@enxebre @vincepri @fabriziopandini WDYT?

Reply:
Moved to: https://github.com/kubernetes-sigs/cluster-api/issues/5332